OSPR-2818 - Add Contact Us form customization options (BB-548)

giovannicimolin opened the pull request:
This adds functionality on LMS to disable the "Contact Us" and to customize the "Contact Us" link with a custom URL. 
To allow for this customization, two SiteConfiguration variables were added:
* CONTACT_US_ENABLE: Switch to enable/disable the Contact Us page. Setting this to
False will disable the contact page and the links on the footer will disappear.
* CONTACT_US_CUSTOM_LINK: If the contact page is enabled, this setting allows to set
a custom URL for the Contact Us links on edx-platform. If this setting is not set,
the platform will use the default contact form.

**JIRA tickets**: [OSPR-2818](https://openedx.atlassian.net/browse/OSPR-2818).

**Dependencies**: None.

**Sandbox URL**: [Sandbox here.](https://pr19315.sandbox.opencraft.hosting/)

**Merge deadline**: None.

**Testing instructions**:

1. Check if the "Contact Us" footer link is working properly
2. Go to django admin -> site configuration and add `"CONTACT_US_ENABLE" = false`
3. Reload django configs and check that the "Contact Us" button disappeared from the LMS as shown on screenshot
4. Go to django admin -> site configuration and change `"CONTACT_US_ENABLE" = true` and `"CONTACT_US_CUSTOM_LINK": "https://open.edx.org/"`
5. Reload django configs and check that the "Contact Us" button leads to https://open.edx.org/

**Reviewers**
- [ ] @xitij2000 
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @giovannicimolin! I've created [OSPR-2818](https://openedx.atlassian.net/browse/OSPR-2818) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

natabene commented:
@giovannicimolin This is great, thank you for your contribution. @marcotuts Do you want to give it a look?

mduboseedx commented:
jenkins ok to test

xitij2000 commented:
jenkins run bokchoy

marcotuts commented:
I believe at a high level this would be helpful, thanks for the contribution. I think we have lots more places where this kind of setting would be useful, I can gather a list of other email addressed but if you wouldn't mind checking for other email addresses in the system it might be helpful to include those locations as well. (Ok to skip that in this PR though if that addition introduces complications).

xitij2000 commented:
@marcotuts This particular change isn't about the contact email though, it's about the /support/contact_us page which relies on Zendesk and as such isn't functional without it. It allows disabling the contact us link or changing it to an alternate contact page. 

We'll check for other places where the /support/contact_us link is used though, and make sure they also use this setting.

natabene commented:
@xitij2000  Do you want to let me know once it is ready?

xitij2000 commented:
@natabene Sure. I think the PR was ready from our POV, but we'll look into uniformly changing all instances of the contact us page with this link (if there are any). I'd love if someone could see if the overall approach here is OK. 

I'll ping you once I have an update on the other links.

giovannicimolin commented:
@xitij2000 and @natabene. 
I've already checked for references to `/support/contact_us` and `support:contact_us` in all files and it doesn't seem like there's any other than the ones from this PR on the code.

natabene commented:
@giovannicimolin Unfortunately, not yet. @ormsbee Any chance you can give this a look?

ormsbee commented on lms/djangoapps/branding/api.py:
```diff
@@ -170,6 +170,47 @@ def _footer_social_links():
+    Returns url of support form, which can have 3 possible values:
+    - '' if the contact form is disabled (by setting SiteConfiguration
+       CONTACT_US_PAGE = True), by the wich is ignored by the
+       parent function
```
Could you please re-phrase this? I'm confused as to what it means.

ormsbee commented on lms/djangoapps/branding/api.py:
```diff
@@ -170,6 +170,47 @@ def _footer_social_links():
+       parent function
+    - The normal edx support form path if CONTACT_US_PAGE is not set
+    - CONTACT_US_PAGE if the the user has set a custom URL redirect
+      for support forms
```
It seems like it could cause problems down the line to have a value that can be either boolean or string. Someone can easily write code to check for `if contact_us:` and think that means "true/false" instead of "true/false/truthy-string". Could we separate it out into two separate config values -- one for whether or not to display a contact page at all, and the other to indicate if a custom link should be used instead of the default? It's a little more verbose, but I think it's less prone to error.

giovannicimolin replied:
That was my first implementation, but I ended up changing it down the line. We decided to use a single variable because this was such a simple feature and using two variables wasn't strictly necessary. 
I can make those changes so we end up with `CONTACT_US_ENABLE` and `CONTACT_US_CUSTOM_LINK`.

ormsbee replied:
Yeah, please do the changes. I realize it's a small thing, but I do think it can confuse folks to have the same variable be very different types. Thank you.

ormsbee commented on lms/djangoapps/branding/api.py:
```diff
@@ -170,6 +170,47 @@ def _footer_social_links():
+            contact_us_page = reverse("support:contact_us")
+
+            if full_path:
+                return '{base_url}{contact_us_path}'.format(
```
Isn't one of the use cases meant to enable entirely external contact forms (e.g. `https://help.edx.org`)? If so, we shouldn't be prepending the LMS base URL, right?

Is it possible to just always echo the URL out as it was entered in configuration, so they could do whatever combination of absolute/relative/fully-qualified link they needed?

giovannicimolin replied:
If we are using external contact forms this piece of code isn't run because of  `if isinstance(contact_us_conf, bool):` a few lines above.

This was implemented because one link on the footer used full path and the other one used relative path and I wanted to keep the current behaviour and not break any tests.

ormsbee commented on lms/djangoapps/branding/api.py:
```diff
@@ -170,6 +170,47 @@ def _footer_social_links():
+        else:
+            return contact_us_conf
+
+    return ''
```
Is this going to eventually generate a blank link (i.e. a link with `<a>` tags but no content inside)? If so, I think that's an accessibility problem, and we'd need to make sure the link was gone entirely in the output.

giovannicimolin replied:
That's treated on the parent function, were all empty links are shown on the page.
[here](https://github.com/edx/edx-platform/blob/ca262ffe0f98a613f1f0a9d13fec45946e92fb87/lms/djangoapps/branding/api.py#L254) and [here](https://github.com/edx/edx-platform/blob/ca262ffe0f98a613f1f0a9d13fec45946e92fb87/lms/djangoapps/branding/api.py#L230)

giovannicimolin commented:
@ormsbee I've addressed your comments and made the necessary changes to the code. I also rebased and squashed the commits and added a more meaningful commit description.

natabene commented:
@ormsbee Are there still changes you requested that have not been made yet?

edx-webhook commented:
@giovannicimolin 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/19315&entry.1484655318=https://github.com/giovannicimolin&entry.752974735=2018-11-26+20:36&entry.1917517419=2019-02-28+01:36&entry.2133058324=Yes) so we can improve your experience in the future.
